moving cliDeleter instance to log.cpp

theSundayProgrammer opened the pull request:
The sequence of deletion has to be different to the reverse of sequence of creation. Hence cliDeleter is necessary.

rcurtin commented:
Hm, do you think instead it would work if we moved the Log objects to cli.cpp?  Then we could remove CLIDeleter entirely because we could guarantee that Log and CLI would be in the same translation unit and therefore control the order of destruction.

It seems like this is the better direction to a solution than #521, so should we close that PR?

theSundayProgrammer commented:
I agree #521 needs to be closed.

theSundayProgrammer commented:
Regarding your other suggestion of moving the log objects to CLI: all that is required to ensure that the singleton CLI object is deleted while the Log object are alive. Moving cliDeleter object to Log.cpp should do that. 
Since ProgramDoc constructor uses CLI singleton and we cannot guarantee that ProgramDoc will be constructed only after the Log objects are created because they are in different files, I am now fixated unless I see code to the contrary, on the the idea that all we can do is ensure  cliDeleter is constructed after Log objects and that no static construction of CLI uses Log objects.

rcurtin commented:
There is... one other possibility.  This could be appealing if the idea of getting all of the classes associated with CLI or mlpack inside of libmlpack.so, and it's also appealing from a code size reduction standpoint.  (I think this also means that if you just `#include <mlpack/core.hpp>` without using anything in it, you don't need to link against -lmlpack.)

The idea is to make ProgramDoc a singleton of its own, then make the PROGRAM_INFO() macro a call to access the ProgramDoc singleton and set its parameters.  The ProgramDoc singleton can then be put into some file like `src/mlpack/core/util/singletons.cpp` which can also contain the CLI and Log objects (technically the Log objects aren't singletons, they're global variables, but hey, I couldn't think of a better name for it...).  Then we can control the order of construction (and thus destruction) and can eliminate the need for CLIDeleter entirely.

What do you think?  Maybe there is something I have overlooked?

Thanks for looking into this, by the way.  The CLI module hasn't gotten much attention since it was written in 2011. :)

theSundayProgrammer commented:
Let me paraphrase what I understood:
----------------------------------------------------
In singleton.cpp you are going to declare
Log Log::info
CLI CLI::singleton
ProgramDoc ProgramDoc::singleton

Then the PROGRAM_INFO macro would  refer to ProgramDoc::singleton directly
or indirectly.
-------------------------------------------------

There is no guarantee that ProgramDoc::singleton would have been created
before the macro which must represent an object is created. Why can't we
just move PROGRAM_INFO et al. inside main. Then we can make sure that
CLI::singleton is created after Log::Info and ilk. I am not sure of the
ramifications though.

I came into CLI_Deleter because I wanted to build a mlpack.lib-free
executable. I removed `#include "cli.hpp"` from core.hpp and managed to run
a simple CNN trainer (basically the mnist test)  on Raspberry Pi without
compiling the mlpack library. I just added log and prefixedoutstream cpp
files.

My suggestion is to move CLI away from core and include it explicitly in
all the applications. This will help better adoption of mlpack.






On Fri, Feb 26, 2016 at 12:40 PM, Ryan Curtin <notifications@github.com>
wrote:

> There is... one other possibility. This could be appealing if the idea of
> getting all of the classes associated with CLI or mlpack inside of
> libmlpack.so, and it's also appealing from a code size reduction
> standpoint. (I think this also means that if you just #include
> <mlpack/core.hpp> without using anything in it, you don't need to link
> against -lmlpack.)
>
> The idea is to make ProgramDoc a singleton of its own, then make the
> PROGRAM_INFO() macro a call to access the ProgramDoc singleton and set its
> parameters. The ProgramDoc singleton can then be put into some file like
> src/mlpack/core/util/singletons.cpp which can also contain the CLI and
> Log objects (technically the Log objects aren't singletons, they're global
> variables, but hey, I couldn't think of a better name for it...). Then we
> can control the order of construction (and thus destruction) and can
> eliminate the need for CLIDeleter entirely.
>
> What do you think? Maybe there is something I have overlooked?
>
> Thanks for looking into this, by the way. The CLI module hasn't gotten
> much attention since it was written in 2011. :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mlpack/mlpack/pull/524#issuecomment-189073562>.
>



-- 
Joseph Chakravarti Mariadassou
http://thesundayprogrammer.com

rcurtin commented:
Okay, I see what you mean and why a header-only library would be useful.  There are some arguments against a header-only library, though: slow compile times, and linking nightmares of the sort that Armadillo has.  Given the significant complexity of mlpack and its use of templates, I would really like to try and keep user compile time low, if possible, instead of having to compile lots and lots of mlpack every time you build something that uses mlpack.

Moving PROGRAM_INFO into main() is fraught with its own problems (specifically, that weird things will happen if you declare PROGRAM_INFO outside of main() and this will eventually confuse users and cause bug reports that have to be dealt with via long-winded explanations later), so I'd like to avoid that.

We can easily guarantee that ProgramDoc::singleton is created before the macro, by creating it during the macro, using a GetSingleton() function like the one you proposed in #521.  Really it's the destruction order that matters more, so that if the CLI interface is being used and -v is specified, the timers are printed correctly.

Another advantage that putting things into a singleton.cpp file has is that then if you are not using the CLI object or the Log object, then I think you won't need to link against mlpack (assuming that you don't have any other functionality that is part of libmlpack.so).  I think that most people aren't using the CLI or Log objects at all in their own code, they are probably just using the algorithms (like, e.g., `mlpack::KMeans<>` or whatever).

rcurtin commented:
@theSundayProgrammer: did you have a chance to take a look into this?  If you don't have time, I can implement a `singletons.cpp` file soon and that should help resolve the original problem.  I'm hoping to release mlpack 2.0.2 soon, and I'm hoping to get this fixed, since you are right that this issue can be a problem.

theSundayProgrammer commented:
HI Ryan
I haven't been able to look at at the issue. I am looking forward to your
solution.
Regards
Joe

On Tue, May 3, 2016 at 1:55 AM, Ryan Curtin <notifications@github.com>
wrote:

> @theSundayProgrammer <https://github.com/theSundayProgrammer>: did you
> have a chance to take a look into this? If you don't have time, I can
> implement a singletons.cpp file soon and that should help resolve the
> original problem. I'm hoping to release mlpack 2.0.2 soon, and I'm hoping
> to get this fixed, since you are right that this issue can be a problem.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/mlpack/mlpack/pull/524#issuecomment-216275406>
>



-- 
Joseph Chakravarti Mariadassou
http://thesundayprogrammer.com

rcurtin commented:
Closing because of #800.
